# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/gtk/nsWindow.cpp
# Commit: 8c7b8e63e417
# Full Hash: 8c7b8e63e417be23932362bc9356f2bbf83f8455
# Author: stransky <stransky@redhat.com>
# Date: 2021-12-20 21:51:27
# Regressor Bug: 1746423
# File Overlap Count: 1
# Description:
#   Bug 1746423 [Linux] Remove static XWindow setup and get it on demand, r=rmader
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D134024
# ==============================================================================

diff -r aacc456d8cd4 -r 8c7b8e63e417 widget/gtk/nsWindow.cpp
--- a/widget/gtk/nsWindow.cpp	Mon Dec 20 10:20:23 2021 +0000
+++ b/widget/gtk/nsWindow.cpp	Mon Dec 20 10:22:19 2021 +0000
@@ -483,13 +483,6 @@
       ,
       mRootAccessible(nullptr)
 #endif
-#ifdef MOZ_X11
-      ,
-      mXWindow(X11None),
-      mXVisual(nullptr),
-      mXDepth(0),
-      mIsShaped(false)
-#endif
 #ifdef MOZ_WAYLAND
       ,
       mNativePointerLockCenter(LayoutDeviceIntPoint()),
@@ -5224,7 +5217,8 @@
   LOG("nsWindow::EnableRenderingToWindow()");
 
   if (mCompositorWidgetDelegate) {
-    mCompositorWidgetDelegate->EnableRendering(mXWindow, mIsShaped);
+    mCompositorWidgetDelegate->EnableRendering(GetX11Window(),
+                                               GetShapedState());
   }
 
   if (GdkIsWaylandDisplay()) {
@@ -5253,23 +5247,35 @@
   }
 }
 
+Window nsWindow::GetX11Window() {
+  return GdkIsX11Display() ? gdk_x11_window_get_xid(mGdkWindow) : X11None;
+}
+
+void nsWindow::EnsureGdkWindow() {
+  if (!mGdkWindow) {
+    mGdkWindow = gtk_widget_get_window(mDrawToContainer ? GTK_WIDGET(mContainer)
+                                                        : mShell);
+    g_object_set_data(G_OBJECT(mGdkWindow), "nsWindow", this);
+  }
+  MOZ_DIAGNOSTIC_ASSERT(mGdkWindow, "We're missing GdkWindow!");
+}
+
+bool nsWindow::GetShapedState() {
+  return mIsTransparent && !mHasAlphaVisual && !mTransparencyBitmapForTitlebar;
+}
+
 void nsWindow::ConfigureGdkWindow() {
   LOG("nsWindow::ConfigureGdkWindow() [%p]", this);
 
-  mGdkWindow =
-      gtk_widget_get_window(mDrawToContainer ? GTK_WIDGET(mContainer) : mShell);
-  g_object_set_data(G_OBJECT(mGdkWindow), "nsWindow", this);
+  EnsureGdkWindow();
 
 #ifdef MOZ_X11
   if (GdkIsX11Display()) {
-    mXWindow = gdk_x11_window_get_xid(mGdkWindow);
-
     GdkVisual* gdkVisual = gdk_window_get_visual(mGdkWindow);
-    mXVisual = gdk_x11_visual_get_xvisual(gdkVisual);
-    mXDepth = gdk_visual_get_depth(gdkVisual);
-    mIsShaped =
-        mIsTransparent && !mHasAlphaVisual && !mTransparencyBitmapForTitlebar;
-    mSurfaceProvider.Initialize(mXWindow, mXVisual, mXDepth, mIsShaped);
+    Visual* visual = gdk_x11_visual_get_xvisual(gdkVisual);
+    int depth = gdk_visual_get_depth(gdkVisual);
+    mSurfaceProvider.Initialize(GetX11Window(), visual, depth,
+                                GetShapedState());
 
     // Set window manager hint to keep fullscreen windows composited.
     //
@@ -5346,15 +5352,6 @@
   }
 
   mSurfaceProvider.CleanupResources();
-
-#ifdef MOZ_X11
-  if (GdkIsX11Display()) {
-    mXWindow = X11None;
-    mXVisual = nullptr;
-    mXDepth = 0;
-    mIsShaped = false;
-  }
-#endif
 }
 
 nsresult nsWindow::Create(nsIWidget* aParent, nsNativeWidget aNativeParent,
@@ -9012,17 +9009,14 @@
   nsCString displayName;
 
   LOG("nsWindow::GetCompositorWidgetInitData");
+  EnsureGdkWindow();
+
+  *aInitData = mozilla::widget::GtkCompositorWidgetInitData(
+      GetX11Window(), displayName, GetShapedState(), GdkIsX11Display(),
+      GetClientSize());
+
 #ifdef MOZ_X11
-  if (GdkIsX11Display() && !mGdkWindow) {
-    LOG("  create mGdkWindow/mXWindow\n");
-    mGdkWindow = gtk_widget_get_window(mDrawToContainer ? GTK_WIDGET(mContainer)
-                                                        : mShell);
-    if (mGdkWindow) {
-      g_object_set_data(G_OBJECT(mGdkWindow), "nsWindow", this);
-      mXWindow = gdk_x11_window_get_xid(mGdkWindow);
-    }
-  }
-  if (GdkIsX11Display() && mXWindow != X11None) {
+  if (GdkIsX11Display()) {
     // Make sure the window XID is propagated to X server, we can fail otherwise
     // in GPU process (Bug 1401634).
     Display* display = DefaultXDisplay();
@@ -9030,9 +9024,6 @@
     displayName = nsCString(XDisplayString(display));
   }
 #endif
-  *aInitData = mozilla::widget::GtkCompositorWidgetInitData(
-      (mXWindow != X11None) ? mXWindow : (uintptr_t) nullptr, displayName,
-      mIsShaped, GdkIsX11Display(), GetClientSize());
 }
 
 #ifdef MOZ_X11